taichi/com_mgr/kernel_compilation_manager.cpp

```diff
@@ -0,0 +1,217 @@
+    const CompileConfig &compile_config,
+    const DeviceCapabilityConfig &caps,
+    const Kernel &kernel_def) const {
+  auto &compiler = *config_.kernel_compiler;
+  auto ir = compiler.compile(compile_config, kernel_def);
+  auto ckd = compiler.compile(compile_config, caps, kernel_def, *ir);
+  return ckd;
+}
+
+std::string KernelCompilationManager::make_kernel_key(
+    const CompileConfig &compile_config,
+    const DeviceCapabilityConfig &caps,
+    const Kernel &kernel_def) const {
+  auto kernel_key = kernel_def.get_cached_kernel_key();
+  if (kernel_key.empty()) {
+    if (!kernel_def.ir_is_ast()) {
+      kernel_key = kernel_def.get_name();
+    } else {  // The kernel key is generated from AST
+      kernel_key = get_hashed_offline_cache_key(compile_config, caps,
+                                                (Kernel *)&kernel_def);
+    }
+
+    kernel_def.set_kernel_key_for_cache(kernel_key);
+  }
+  return kernel_key;
+}
+
+const CompiledKernelData *KernelCompilationManager::try_load_cached_kernel(
+    const Kernel &kernel_def,
+    const std::string &kernel_key,
+    Arch arch,
+    CacheData::CacheMode cache_mode) {
+  {  // Find in memory-cache (caching_kernels_)
+    const auto &kernels = caching_kernels_;
+    auto iter = kernels.find(kernel_key);
+    if (iter != kernels.end()) {
+      TI_DEBUG("Create kernel '{}' from in-memory cache (key='{}')",
+               kernel_def.get_name(), kernel_key);
+      return iter->second.compiled_kernel_data.get();
+    }
+  }
+  // Find in disk-cache (cached_data_)
+  if (cache_mode == CacheData::MemAndDiskCache) {
+    auto &kernels = cached_data_.kernels;
+    auto iter = kernels.find(kernel_key);
+    if (iter != kernels.end()) {
+      auto &k = iter->second;
+      if (auto loaded = load_ckd(kernel_key, arch)) {
+        TI_DEBUG("Create kernel '{}' from cache (key='{}')",
+                 kernel_def.get_name(), kernel_key);
+        TI_ASSERT(loaded->arch() == arch);
+        k.last_used_at = std::time(nullptr);
+        k.compiled_kernel_data = std::move(loaded);
+        updated_data_.push_back(&k);
+        return k.compiled_kernel_data.get();
+      }
+    }
+  }
+  return nullptr;
+}
+
+const CompiledKernelData &KernelCompilationManager::compile_and_cache_kernel(
+    const std::string &kernel_key,
+    const CompileConfig &compile_config,
+    const DeviceCapabilityConfig &caps,
+    const Kernel &kernel_def) {
+  auto cache_mode = get_cache_mode(compile_config, kernel_def);
+  TI_DEBUG_IF(cache_mode == CacheData::MemAndDiskCache,
+              "Cache kernel '{}' (key='{}')", kernel_def.get_name(),
+              kernel_key);
+  TI_ASSERT(caching_kernels_.find(kernel_key) == caching_kernels_.end());
+  KernelCacheData k;
+  k.kernel_key = kernel_key;
+  k.created_at = k.last_used_at = std::time(nullptr);
+  k.compiled_kernel_data = compile_kernel(compile_config, caps, kernel_def);
+  k.size = k.compiled_kernel_data->size();
```

Comment:
nit: shall we just do compile and call another function "add_to_cache" method so that they're separate? ;)

Reply:
Maybe we can provide more detailed APIs:
(possible function set)
* `KernelCompilationManager::load_or_compile`: To compile && auto-cache
* `KernelCompilationManager::compile`: To compile without caching
* `KernelCompilationManager::load_from_cache`: To load compilation result from cache manualy
* `KernelCompilationManager::update_cache`(or `add_to_cache`): To add/update the compilation result to cache manually

Reply:
You mean we should provide more detailed APIs?

(possible function set)
* `KernelCompilationManager::load_or_compile`: To compile && auto-cache
* `KernelCompilationManager::compile`: To compile without caching
* `KernelCompilationManager::load_from_cache`: To load compilation result from cache manualy
* `KernelCompilationManager::update_cache`(or `add_to_cache`): To add/update the compilation result to cache manually

Reply:
Yup exactly, the update_cache part definitely worths a modular function :P 
super super nit: To match `load_from_cache` we probably prefer `save_to_cache` ;)

Reply:
OK. I will do it in next PRs. : )